Terraform Crash dynamic tags

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!--- 0
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->
.12

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

variables.tf:

```hcl

variable "SpotFleetInstanceTags" {
  type        = map
  description = "EC2 Instance Tags"
}
```

main.tf:
```hcl
resource "aws_spot_fleet_request" "xxxSpotFleet" {
  iam_fleet_role      = "${var.SpotFleetIamFleetRole}"
  spot_price          = "0.03"
  allocation_strategy = "diversified"
  target_capacity     = 2
  valid_until         = "2019-11-04T20:44:20Z"
  dynamic "launch_specification" {
    iterator = instanceType
    for_each = var.SpotFleetInstanceTypes
    content {
         instance_type            = instanceType.value
         ami                      = "${var.SpotFleetimageId}"
         key_name                 = "my-key"
         spot_price               = "${var.SpotFleetPrice}"
         iam_instance_profile_arn = "${var.SpotFleetIamInstanceProfile}"
         subnet_id                = "${var.SpotFleetSubnetId}"
         weighted_capacity        = "${var.SpotFleetWeightedCapacity}"
         dynamic "tags" {
             for_each = var.SpotFleetInstanceTags
             iterator = fleetTag
             content {
               key   = fleetTag.key
               value = fleetTag.value
               propagate_at_launch = true
             }
         }
   }
  }
}
```

vars.tfvars:

```hcl
SpotFleetInstanceTags = {
    Name = "Create-Compute-SpotFleet-all-xxx-xxx"
    Environment = "production"
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->

```
panic: AttributeTypes on non-object Type

goroutine 14 [running]:
github.com/zclconf/go-cty/cty.Type.AttributeTypes(...)
        /opt/teamcity-agent/work/9e329aa031982669/pkg/mod/github.com/zclconf/go-cty@v1.0.0/cty/object_type.go:134
github.com/hashicorp/terraform/lang/blocktoattr.SchemaForCtyElementType(0x2206c00, 0xc0000404e0, 0x2206c00)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/lang/blocktoattr/schema.go:108 +0x24d
github.com/hashicorp/terraform/lang/blocktoattr.walkVariables(0x2205880, 0xc0003db4a0, 0xc000484660, 0x1, 0x2205880, 0xc0003db4a0, 0xc00116d080, 0x2, 0xc0011a3b00, 0x1)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:37 +0x476
github.com/hashicorp/terraform/lang/blocktoattr.walkVariables(0x2205880, 0xc0002d9130, 0x0, 0x1, 0x2205880, 0xc0002d9130, 0xc00113ee50, 0x0, 0x0, 0x3311ca0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:35 +0x289
github.com/hashicorp/terraform/lang/blocktoattr.ExpandedVariables(...)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:24
github.com/hashicorp/terraform/lang.ReferencesInBlock(0x2205880, 0xc0002d9130, 0xc00113ee50, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc000450d80)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/lang/references.go:68 +0x66
github.com/hashicorp/terraform/terraform.(*NodeAbstractResource).References(0xc0000bc2c0, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/node_resource_abstract.go:190 +0x251
github.com/hashicorp/terraform/terraform.NewReferenceMap(0xc000470960, 0xf, 0xf, 0xf)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/transform_reference.go:402 +0x24a
github.com/hashicorp/terraform/terraform.(*ReferenceTransformer).Transform(0x3311ca0, 0xc00009aaf0, 0x14, 0x0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/transform_reference.go:73 +0x71
github.com/hashicorp/terraform/terraform.(*BasicGraphBuilder).Build(0xc0004ef8e0, 0x0, 0x0, 0x0, 0x32f1fe0, 0xc0004ef910, 0x4e5650, 0xc00009a000)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/graph_builder.go:50 +0x477
github.com/hashicorp/terraform/terraform.(*PlanGraphBuilder).Build(0xc0000ecf50, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1a7b780, 0xc0002f2cc0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/graph_builder_plan.go:65 +0xbc
github.com/hashicorp/terraform/terraform.(*Context).Graph(0xc000e2e000, 0x6, 0xc0004ef966, 0xc0004efa00, 0x1, 0x1, 0x0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/context.go:292 +0x1b8
github.com/hashicorp/terraform/terraform.(*Context).Validate(0xc000e2e000, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/terraform/context.go:670 +0xf3
github.com/hashicorp/terraform/backend/local.(*Local).context(0xc0004cc270, 0xc0004a1180, 0x1, 0x1, 0x2, 0xc00000c100, 0x1, 0x2, 0x0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend_local.go:120 +0x974
github.com/hashicorp/terraform/backend/local.(*Local).opPlan(0xc0004cc270, 0x22051c0, 0xc000084280, 0x22051c0, 0xc000084740, 0xc0004a1180, 0xc0000841c0)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend_plan.go:67 +0x2a3
github.com/hashicorp/terraform/backend/local.(*Local).Operation.func1(0xc0000825a0, 0xc000082710, 0xc000082a70, 0xc0004a1180, 0xc0004cc270, 0xc0000841c0, 0xc000082000, 0x22051c0, 0xc000084280, 0x22051c0, ...)
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend.go:356 +0x135
created by github.com/hashicorp/terraform/backend/local.(*Local).Operation
        /opt/teamcity-agent/work/9e329aa031982669/src/github.com/hashicorp/terraform/backend/local/backend.go:340 +0x36b
!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Expected Behavior
<!--
What should have happened?
-->
Just add tags to spot fleet.

### Actual Behavior
<!--
What actually happened?
-->
Crashed
### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
I just need terraform plan --var-file="vars.tfvars"
every run even if I delete the state files gives same error.


### Additional Context
<!--
Are there anything atypical about your situation that we should know? For example: is Terraform running in a wrapper script or in a CI system? Are you passing any unusual command line options or environment variables to opt-in to non-default behavior?
-->
the only feeling I have it Dynamic loop inside another dynamic loop. 
in the code I am running thru instance types using dynamic and then again dynamics for tags.  Code was running until I change the standard tag syntax to Dynamic tags. 
### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- #6017

-->

--- a/lang/blocktoattr/variables.go
+++ b/lang/blocktoattr/variables.go
@@ -33,7 +33,7 @@ func walkVariables(node dynblock.WalkVariablesNode, body hcl.Body, schema *confi
 	for _, child := range children {
 		if blockS, exists := schema.BlockTypes[child.BlockTypeName]; exists {
 			vars = append(vars, walkVariables(child.Node, child.Body(), &blockS.Block)...)
-		} else if attrS, exists := schema.Attributes[child.BlockTypeName]; exists {
+		} else if attrS, exists := schema.Attributes[child.BlockTypeName]; exists && attrS.Type.ElementType().IsObjectType() {
 			synthSchema := SchemaForCtyElementType(attrS.Type.ElementType())
 			vars = append(vars, walkVariables(child.Node, child.Body(), synthSchema)...)
 		}

--- a/lang/blocktoattr/variables_test.go
+++ b/lang/blocktoattr/variables_test.go
@@ -21,6 +21,10 @@ func TestExpandedVariables(t *testing.T) {
 				})),
 				Optional: true,
 			},
+			"bar": {
+				Type:     cty.Map(cty.String),
+				Optional: true,
+			},
 		},
 	}
 
@@ -138,6 +142,29 @@ dynamic "foo" {
 							Filename: "test.tf",
 							Start:    hcl.Pos{Line: 5, Column: 11, Byte: 57},
 							End:      hcl.Pos{Line: 5, Column: 14, Byte: 60},
+						},
+					},
+				},
+			},
+		},
+		"misplaced dynamic block": {
+			src: `
+dynamic "bar" {
+  for_each = beep
+  content {
+    key = val
+  }
+}
+`,
+			schema: fooSchema,
+			want: []hcl.Traversal{
+				{
+					hcl.TraverseRoot{
+						Name: "beep",
+						SrcRange: hcl.Range{
+							Filename: "test.tf",
+							Start:    hcl.Pos{Line: 3, Column: 14, Byte: 30},
+							End:      hcl.Pos{Line: 3, Column: 18, Byte: 34},
 						},
 					},
 				},

